contrib/epee/src/net_ssl.cpp

```diff
@@ -519,10 +520,16 @@ bool ssl_options_t::handshake(
 
   boost::system::error_code ec = boost::asio::error::would_block;
   socket.async_handshake(type, boost::lambda::var(ec) = boost::lambda::_1);
-  while (ec == boost::asio::error::would_block)
+  if (io_service.stopped())
   {
     io_service.reset();
-    io_service.run_one();
+  }
+  while (ec == boost::asio::error::would_block)
+  {
+    // should poll_one(), can't run_one() because it can block if there is
+    // another worker thread executing io_service's tasks
+    std::this_thread::sleep_for(std::chrono::milliseconds(30));
```

Comment:
Add to the comment that boost 1.66+ has `run_one_for` which does both in one call. It should be more efficient if asio uses the system calls properly.

Reply:
I know. But monero uses boost 1.58

Reply:
Added a TODO